### PR TITLE
fix(postgres): sync with alter method fails with dataType enum (#7649)

### LINKED ISSUE #7649: sync() with new alter method fails with dataType enum
## What you are doing?

```js
'use strict';

const Sequelize = require('sequelize');

//first connect to the database
const database = new Sequelize('database', 'user', 'password', {
    dialect: 'postgres',
    host: '127.0.0.1',
    port: 5432,

    pool: {
        max: 5,
        min: 0,
        idle: 10000
    },
});

return database.authenticate()
.then(() => {
    //success connecting,
    console.log('database connected');
    return database;
})
.then(database => {
    const containerModel = database.define('TestContainer', {
        option: {
            type: Sequelize.DataTypes.ENUM('enum1', 'enum2', 'enum3'),
            defaultValue: 'enum1',
            allowNull: false
        },
    });

    return database.sync({
        alter: true,
        force: false,
    });
})
.catch(err => {
    //failed to connect
    console.error(err);
    process.exit(1);
});
```

## What do you expect to happen?
Being able to run the example multiple times without fail

## What is actually happening?
On second run, an error occurs:

> SequelizeDatabaseError: type "enum_TestContainer_option" already exists

__Dialect:__ postgres
__Database version:__ 9.6.2
__Sequelize version:__ git master

### PR DESCRIPTION
**Pull request by aristov**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Closes #7649 

This change fixes an error that occurs when calling `sync({ alter : true })` on a model with the ENUM data type.

## Todos

- [ ] This only fixes the sync() failure, but sync can't delete/replace the enum value yet

### CODE REVIEW

**Review by ephys — CHANGES_REQUESTED** [review]
Thank you for the PR :)

File: src/dialects/postgres/query-generator.js [hunk 1]
```diff
@@ -783,7 +783,7 @@ class PostgresQueryGenerator extends AbstractQueryGenerator {
       values = dataType.toString().match(/^ENUM\(.+\)/)[0];
     }
 
-    let sql = `CREATE TYPE ${enumName} AS ${values};`;
+    let sql = `DO $$ BEGIN CREATE TYPE ${enumName} AS ${values}; EXCEPTION WHEN duplicate_object THEN null; END $$;`;
```
[the remark this answers is not in the document]
Comment (aristov) [reply src/dialects/postgres/query-generator.js:786]: > The contents of the DO block can use a regular single-quote string

But what about enum values? They should also be enclosed in regular single-quotes.

[the remark this answers is not in the document]
Comment (aristov) [reply src/dialects/postgres/query-generator.js]: Done.

File: test/unit/sql/enum.test.js [hunk 2]
```diff
@@ -43,13 +43,13 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
       describe('pgEnum', () => {
         it('uses schema #3171', () => {
           expectsql(sql.pgEnum(FooUser.getTableName(), 'mood', FooUser.rawAttributes.mood.type), {
-            postgres: 'CREATE TYPE "foo"."enum_users_mood" AS ENUM(\'happy\', \'sad\');'
+            postgres: 'DO \'BEGIN CREATE TYPE "foo"."enum_users_mood" AS ENUM(\'\'happy\'\', \'\'sad\'\'); EXCEPTION WHEN duplicate_object THEN null; END\';'
```
Comment (WikiRik) [inline test/unit/sql/enum.test.js:46]:
```suggestion
            postgres: `DO 'BEGIN CREATE TYPE "foo"."enum_users_mood" AS ENUM(''happy'', ''sad''); EXCEPTION WHEN duplicate_object THEN null; END';`
```
This should work, although the double single quotes in the enum is strange

File: test/unit/sql/enum.test.js [hunk 3]
```diff
@@ -43,13 +43,13 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
...
           });
         });
 
         it('does add schema when public', () => {
           expectsql(sql.pgEnum(PublicUser.getTableName(), 'theirMood', PublicUser.rawAttributes.mood.type), {
-            postgres: 'CREATE TYPE "public"."enum_users_theirMood" AS ENUM(\'happy\', \'sad\');'
+            postgres: 'DO \'BEGIN CREATE TYPE "public"."enum_users_theirMood" AS ENUM(\'\'happy\'\', \'\'sad\'\'); EXCEPTION WHEN duplicate_object THEN null; END\';'
```
Comment (WikiRik) [inline test/unit/sql/enum.test.js:52]:
```suggestion
            postgres: `DO 'BEGIN CREATE TYPE "public"."enum_users_theirMood" AS ENUM(''happy'', ''sad''); EXCEPTION WHEN duplicate_object THEN null; END';`
```

File: test/unit/dialects/postgres/query-generator.test.js [hunk 4]
```diff
@@ -306,7 +306,7 @@ if (dialect.startsWith('postgres')) {
...
+          expectation: 'ALTER TABLE "myTable" ALTER COLUMN "col_1" SET NOT NULL;ALTER TABLE "myTable" ALTER COLUMN "col_1" DROP DEFAULT;DO \'BEGIN CREATE TYPE "public"."enum_myTable_col_1" AS ENUM(\'\'value 1\'\', \'\'value 2\'\'); EXCEPTION WHEN duplicate_object THEN null; END\';ALTER TABLE "myTable" ALTER COLUMN "col_1" TYPE "public"."enum_myTable_col_1" USING ("col_1"::"public"."enum_myTable_col_1");ALTER TABLE "myTable" ALTER COLUMN "col_2" SET NOT NULL;ALTER TABLE "myTable" ALTER COLUMN "col_2" DROP DEFAULT;DO \'BEGIN CREATE TYPE "public"."enum_myTable_col_2" AS ENUM(\'\'value 3\'\', \'\'value 4\'\'); EXCEPTION WHEN duplicate_object THEN null; END\';ALTER TABLE "myTable" ALTER COLUMN "col_2" TYPE "public"."enum_myTable_col_2" USING ("col_2"::"public"."enum_myTable_col_2");'
```
Comment (WikiRik) [inline test/unit/dialects/postgres/query-generator.test.js:309]:
```suggestion
          expectation: `ALTER TABLE "myTable" ALTER COLUMN "col_1" SET NOT NULL;ALTER TABLE "myTable" ALTER COLUMN "col_1" DROP DEFAULT;DO 'BEGIN CREATE TYPE "public"."enum_myTable_col_1" AS ENUM(''value 1'', ''value 2''); EXCEPTION WHEN duplicate_object THEN null; END';ALTER TABLE "myTable" ALTER COLUMN "col_1" TYPE "public"."enum_myTable_col_1" USING ("col_1"::"public"."enum_myTable_col_1");ALTER TABLE "myTable" ALTER COLUMN "col_2" SET NOT NULL;ALTER TABLE "myTable" ALTER COLUMN "col_2" DROP DEFAULT;DO 'BEGIN CREATE TYPE "public"."enum_myTable_col_2" AS ENUM(''value 3'', ''value 4''); EXCEPTION WHEN duplicate_object THEN null; END';ALTER TABLE "myTable" ALTER COLUMN "col_2" TYPE "public"."enum_myTable_col_2" USING ("col_2"::"public"."enum_myTable_col_2");`
```

File: test/integration/dialects/postgres/sync-alter-enum.test.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+const chai = require('chai'),
+  assert = chai.assert,
+  Support = require('../../support'),
+  Sequelize = Support.Sequelize,
+  dialect = Support.getTestDialect();
+
+if (dialect.match(/^postgres/)) {
+  describe('[POSTGRES Specific] sync with alter method with dataType enum', () => {
+    it('properly sync, #7649', async function() {
```
Comment (WikiRik) [inline test/integration/dialects/postgres/sync-alter-enum.test.js:11]: To reduce nesting can you add something like the following here?

```
if (dialect !== 'postgres') {
   return;
```
(`getTestDialect` normalizes `postgres-native` to `postgres` iirc)

**Review by WikiRik — COMMENTED** [review]
Few nitpicky suggestions

**Comment by ephys** [comment]
Sorry to ask this but could you open the same PR against the main branch before we merge this one?

If we don't merge it against main, we'll lose this fix in v7

File: test/unit/sql/enum.test.js (hunk 2 above)
Comment (aristov) [reply test/unit/sql/enum.test.js:46]: There is a [problem with ESLint](https://github.com/sequelize/sequelize/blob/d3f5b5a65e297f4b6861e6a6ce335a9830b28781/.eslintrc.json#L67)

File: test/integration/dialects/postgres/sync-alter-enum.test.js (hunk 5 above)
Comment (aristov) [reply test/integration/dialects/postgres/sync-alter-enum.test.js]: Done.

File: test/unit/sql/enum.test.js (hunk 2 above)
Comment (WikiRik) [reply test/unit/sql/enum.test.js:46]: Ah, that has been improved in `main` then. You can ignore this for this PR and only do it for the PR targeting `main`

**Comment by aristov** [comment]
> Sorry to ask this but could you open the same PR against the main branch before we merge this one?
> 
> If we don't merge it against main, we'll lose this fix in v7

https://github.com/sequelize/sequelize/pull/15738

File: test/unit/sql/enum.test.js (hunk 2 above)
Comment (aristov) [reply test/unit/sql/enum.test.js:46]: Ok, it's done: https://github.com/sequelize/sequelize/pull/15738

**Comment by WikiRik** [comment]
I think a few suggestions from #15738 might also be applicable here, if you update this PR we can review it again and land it in the next `v6` release

**Comment by aristov** [comment]
> I think a few suggestions from #15738 might also be applicable here, if you update this PR we can review it again and land it in the next `v6` release

Done.

**Review by ephys — APPROVED** [review]

**Review by WikiRik — APPROVED** [review]

**Pull request merged**